--- samba-4.0.2-orig/dynconfig/wscript.orig	2013-08-26 02:17:02.704395990 -0600
+++ samba-4.0.2/dynconfig/wscript	2013-08-26 02:18:09.361370037 -0600
@@ -156,6 +156,13 @@
          'OPTION':    '--with-pammodulesdir',
          'HELPTEXT':  'Which directory to use for PAM modules',
     },
+    'SYSROOTDIR' : {
+         'STD-PATH':  '${SYSROOTDIR}',
+         'FHS-PATH':  '${SYSROOTDIR}',
+         'OPTION':    '--with-sysrootdir',
+         'HELPTEXT':  'Target sysroot directory (if cross-compiling)',
+         'OVERWRITE': True,
+    },
     'CONFIGDIR' : {
          'STD-PATH':  '${SYSCONFDIR}',
          'FHS-PATH':  '${SYSCONFDIR}/samba',
--- samba-4.0.2-orig/buildtools/wafsamba/wscript.orig	2013-08-26 02:37:13.534217107 -0600
+++ samba-4.0.2/buildtools/wafsamba/wscript	2013-08-26 02:38:27.811316808 -0600
@@ -74,6 +74,10 @@
                    help=("private library directory [PREFIX/lib/%s]" % Utils.g_module.APPNAME),
                    action="store", dest='PRIVATELIBDIR', default=None)
 
+    opt.add_option('--with-sysrootdir',
+                   help=("SYSROOT target directory for cross-compiling"),
+                   action="store", dest='SYSROOTDIR', default=None)
+
     opt.add_option('--with-libiconv',
                    help='additional directory to search for libiconv',
                    action='store', dest='iconv_open', default='/usr/local',
@@ -242,6 +246,7 @@
     conf.env.SBINDIR = Options.options.SBINDIR or '${PREFIX}/sbin'
     conf.env.MODULESDIR = Options.options.MODULESDIR
     conf.env.PRIVATELIBDIR = Options.options.PRIVATELIBDIR
+    conf.env.SYSROOTDIR = Options.options.SYSROOTDIR
     conf.env.BUNDLED_LIBS = Options.options.BUNDLED_LIBS.split(',')
     conf.env.PRIVATE_LIBS = Options.options.PRIVATE_LIBS.split(',')
     conf.env.BUILTIN_LIBRARIES = Options.options.BUILTIN_LIBRARIES.split(',')
